src/util/buffvec.h

```diff
@@ -0,0 +1,893 @@
+    }
+
+    template<typename U = value_type>
+    typename std::enable_if<std::is_trivially_move_constructible<typename std::remove_cv<U>::type>::value && std::is_trivially_destructible<typename std::remove_cv<U>::type>::value>::type
+    reallocate(size_type const new_capacity) {
+        SASSERT(new_capacity >= size());
+        std::size_t const new_bytesize = static_cast<std::size_t>(new_capacity) * sizeof(value_type);
+
+        if(m_data != reinterpret_cast<pointer>(&m_initial_buffer)) {
+            m_data = reinterpret_cast<pointer>(memory::reallocate(m_data, new_bytesize));
+        } else {
+            pointer const buffvec = reinterpret_cast<pointer>(memory::allocate(new_bytesize));
+            buffvec_detail::move_into(buffvec, m_data, m_size);
+            m_data = buffvec;
+        }
+        m_capacity = new_capacity;
+    }
+
+    template<typename U = value_type>
+    typename std::enable_if<!std::is_trivially_move_constructible<typename std::remove_cv<U>::type>::value || !std::is_trivially_destructible<typename std::remove_cv<U>::type>::value>::type
+    reallocate(size_type const new_capacity) {
+        SASSERT(new_capacity >= size());
+        std::size_t const new_bytesize = static_cast<std::size_t>(new_capacity) * sizeof(value_type);
+        pointer const buffvec = reinterpret_cast<pointer>(memory::allocate(new_bytesize));
+
+        buffvec_detail::move_into(buffvec, m_data, m_size);
+        buffvec_detail::destroy(m_data, m_data + m_size);
+        if(m_data != reinterpret_cast<pointer>(&m_initial_buffer)) {
+            memory::deallocate(m_data);
+        }
+
+        m_data = buffvec;
+        m_capacity = new_capacity;
+    }
+
+public:
+    constexpr buffvec() noexcept = default;
+
+    buffvec(buffvec const& other) {
+        size_type const pos = other.m_size;
+        m_size = pos;
+        if(pos <= initial_size) {
+            m_data = reinterpret_cast<pointer>(&m_initial_buffer);
+            m_capacity = initial_size;
+        } else {
+            m_data = reinterpret_cast<pointer>(memory::allocate(static_cast<std::size_t>(pos) * sizeof(value_type)));
+            m_capacity = pos;
+        }
+        buffvec_detail::copy_into(m_data, other.m_data, pos);
+    }
+
+    buffvec& operator=(buffvec const& other) {
+        if(this != &other) {
+            clear();
+            reserve(other.m_size);
+            m_size = other.m_size;
+            buffvec_detail::copy_into(m_data, other.m_data, m_size);
+        }
+        return *this;
+    }
+
+    // Depending on initial_size, move construction may still be very expensive
+    buffvec(buffvec&& other) {
+        if(other.m_data != reinterpret_cast<pointer>(&other.m_initial_buffer)) {
+            m_data = other.m_data;
+            m_size = other.m_size;
+            m_capacity = other.m_capacity;
+            other.m_data = reinterpret_cast<pointer>(&other.m_initial_buffer);
+            other.m_size = 0;
+            other.m_size = initial_size;
```

Comment:
why assigned twice?

Reply:
because the second assignment was supposed to be `other.m_capacity = initial_size;`